core/src/main/scala/kafka/tools/ConsoleProducer.scala

```diff
@@ -362,11 +375,21 @@ object ConsoleProducer {
     }
 
     private def splitHeaders(headers: String): Array[(String, Array[Byte])] = {
-      headerSeparatorPattern.split(headers).map { pair =>
-        (pair.indexOf(headerKeySeparator), ignoreError) match {
+      headersSeparatorPattern.split(headers).map { pair =>
+        (pair.indexOf(headersKeySeparator), ignoreError) match {
           case (-1, false) => throw new KafkaException(s"No header key separator found in pair '$pair' on line number $lineNumber")
           case (-1, true) => (pair, null)
-          case (i, _) => (pair.substring(0, i), pair.substring(i + headerKeySeparator.length).getBytes(StandardCharsets.UTF_8))
+          case (i, _) =>
+            val headerKey = pair.substring(0, i) match {
+              case k =>
+                if (k == nullMarker)
+                  throw new KafkaException(s"Header keys should not be equal to the null marker '$nullMarker' as they can't be null")
+                else k
```

Comment:
nit: We could use `case k if (k == nullMarker) =>` followed by `case k =>` for the default case. That seems a bit more intuitive when using pattern matching.